piksi_tools/testing/usrp_test.sh

```diff
@@ -0,0 +1,207 @@
+    echo "`egrep -o "[^=]+$" FULL_VERSION`" | tr " " "\n" | sed 's/:/: /' > version.yaml
+}
+
+function check_device_settings () {
+    # After the devices have been programmed, asserts that the devices
+    # have been programmed with the right firmware and nap versions by
+    # reading their settings.
+    export HDL_`cat HDL_VERSION`
+    cd $CWD
+    echo $2
+    timeout $BOOTLOAD_TIMEOUT python piksi_tools/diagnostics.py -p $1 -o $2
+    cat version.yaml
+    cd $TEST
+    ../../$VENV_PATH/python check_device_details.py -d ../../$2 -v ../../version.yaml
+    if [[ $? -eq 1 ]]
+    then
+        echo "Check device details failed!"
+    fi
+}
+
+function start_scenario () {
+    # Invoke a USRP testing scenario from the USRP machine.
+    set +x
+    ssh jenkins@nala timeout $PLAYBACK_DURATION \
+        python -u ~henry/swift/peregrine/peregrine/stream_usrp.py -1 -u \
+        name=b200_{1,2} \
+        -g 30 \
+        -p \
+        /data/sky/{leica,novatel}-20150707-070000.1bit
+}
+
+function start_recording () {
+    # Start a serial link SBP recording.
+    cd $CWD
+    echo "Starting recording on" $1
+    echo "Writing output to " $2
+    python piksi_tools/serial_link.py -l -r -t $PLAYBACK_DURATION -p $1 -o $2 > /dev/null &
+}
+
+function prep_logs () {
+    # Create a per-device session logs and cleanup stuff if it's
+    # already there.
+    mkdir -p $3
+    rm -f $3/*
+}
+
+## Running stuff in parallel
+
+function run_parallel_pairs () {
+    # Run a function on each device in a pair, parallel across pairs
+    # (rover + base station).
+    while read DEV1 DEV2; do
+      ROVER_DEV_ID=$(echo $DEV1 | egrep -o 'PK[0-9]+')
+      $1 $DEV1 $DEV2 $ROVER_DEV_ID &
+    done < $DEVICES
+    wait
+}
+
+function program_pair () {
+    echo "Programming firmware on: " $1 $2
+    program_device $1 $3/rover.log
+    program_device $2 $3/base.log
+    sleep 1
+}
+
+function diagnostics_pair () {
+    echo "Check diagnostics on: " $1 $2
+    check_device_settings $1 $3/diagnostics_rover.yaml
+    check_device_settings $2 $3/diagnostics_base.yaml
+}
+
+function record_pair () {
+    echo "Records SBP log data on:" $1 $2
+    start_recording $1 $3/rover.json.log
+    start_recording $2 $3/base.json.log
+}
+
+function run_devices () {
+    # Primary runner across all devices: prepare the log directory,
+    # program pairs of Piksi's (rovers and base stations
+    # sequentially), check that devices were programmed with the right
+    # firmware, kickoff USRP recording and timeout eventually.
+    run_parallel_pairs prep_logs
+    run_parallel_pairs program_pair
+    # Give Piksi a rest for a bit after programming.
+    sleep 10
+    run_parallel_pairs diagnostics_pair
+    run_parallel_pairs record_pair
+    start_scenario
+    echo "Done!"
+}
+
+function show_help () {
+    echo "Move along, nothing to see here..."
+}
+
+while getopts ":x:" opt; do
```

Comment:
If I broke into two parts (1. program., 2., run serial capture, etc.) would it make more sense then?